text/0000-reloadable-process-types.md

```diff
@@ -0,0 +1,36 @@
+# Reloadable Process Types
+
+## Summary
+
+Provide a utility buildpack to support language buildpacks in creating reloadable process types for their run containers.
+
+## Motivation
+
+Usage of dev orchestration tools like Tilt and Skaffold are on the rise. These tools aim to make  developing distributed software that runs on Kubernetes easier.  Offering the capability to develop inside the container via live reload features.  With this, changes are streamed from the developer’s workstation into the running container whereupon the init process is restarted.  The dev orchestrator, of course, handles most of this by building the changes and copying them in the container but ultimately the running container needs the ability to restart its entrypoint process for the changes to take effect.
+
+Today buildpacks are unable to participate in these live reload development workflows but, moving forward, we would very much like them to be able to.
+
+## Detailed Explanation
+
+We propose creating a utility buildpack that packages the watchexec command line utility for restarting arbitrary processes.
+
+When enabled by a BP_LIVE_RELOAD_ENABLED=true environment variable, the language buildpacks may request the inclusion of watchexec, via the build plan.
+
+This affords language buildpacks the ability to use watchexec for creating a reloadable process type when no better options are available.  For example, the java buildpack may prefer to use spring-boot-devtools if that is available but can always fallback to watchexec otherwise.  
+
+As a result, a language buildpack may produce a reloadable process type such as:
+
+`watchexec -r java org.springframework.boot.loader.JarLauncher`
+
+and in doing so, for the reloadable process type, the buildpack makes watchexec the stable init process for the container that monitors the workspace directory for file changes, restarting the app process when they occur.
```

Comment:
> that monitors the workspace directory for file changes, restarting the app process when they occur.

Many buildpacks may also require that you rebuild the source code again before it is run. E.g. Changes to C# source code when used with dotnet-core buildpack. Is this buildpack intended just for Java buildpacks?

Reply:
Correct me if I'm wrong Paul, but I believe this is the tool's responsibility (Tilt/Skaffold).

>The dev orchestrator, of course, handles most of this by building the changes and copying them in the container

Reply:
When I was exploring the potential to use Tilt + buildpacks, one thing I noticed was that a user generally determines the build command(s) that are used to produce the contents of their image. (Often via a Dockerfile). When doing a hot reload, Tilt rebuilds the app image contents using (a subset of) those same build command(s) each time. Then, the resulting artifacts are copied into the running app container.

Buildpack builds have a key difference: buildpacks - not users - determine the build commands that produce the image. And these commands aren't checked into a file – they're implicit in the buildpacks' source code. If rebuilds are delegated to the dev tool (e.g. Tilt), if it doesn't follow the same process as the buildpack's build, you could easily end up with a built artifact that is incompatible with the app container (e.g. built for wrong operating system), or just substantively different from the contents produced by the buildpack (e.g. built with a different set of dependencies).

This could lead to failing (re)builds, containers whose behaviour in development diverges from their behaviour in prod, and other issues.

Are there any ideas about how to mitigate this problem / create an "it just works" experience here?

Reply:
The intention here, I think, is that pack (and this utility buildpack) would be used with `tilt` and `skaffold` custom builds, something like this (tilt example):

```
custom_build('localhost:5000/apps/demo', 
  'pack build -e BP_LIVE_RELOAD_ENABLED=true --tag $EXPECTED_REF --publish,
  ['pom.xml'],
  live_update = [
    sync('./target/classes', '/workspace/BOOT-INF/classes')
  ]
)
k8s_yaml('./k8s/deployment.yaml'))
k8s_resource('spring-petclinic', port_forwards="8080:8080”)
````

When `tilt up` is run this will perform an initial build and deployment, using pack, that publishes an image using the tag TIlt's wants.  It then live sync's a limited set of changes, in this case changes within the existing Java classpath compiled by the IDE, directly into the application container.  Other changes outside of this, to the pom for example, would trigger a full build and deployment again that would again delegate to pack.

Does that clarify things?

Reply:
Done.  See latest commit.

Reply:
The intent here is to make some progress towards the idea of "hot-reloadable" applications using buildpacks, while not requiring that we undertake the entire workstream that would include significant changes upstream to the buildpack spec/lifecycle. The tradeoffs to correctness or ease-of-use are understood, but could maybe be explicitly outlined here so that others reading this can understand this RFCs shortcomings. @paulcwarren, do you think it would be appropriate to include a bit about some of these tradeoffs in the design of this RFC?

Reply:
The tradeoffs being?

Reply:
I think we only have that one concern. It feels worth noting that once the container is initially built, the only thing it can do is swap some files on the filesystem and restart the process. So, if your application needs to be recompiled, this must happen outside of the built container and in so doing must happen in a way that remains compatible with the built container environment.

Reply:
Noted.  Thanks.  I'll update the RFC tomorrow AM.  Thankyou.

Reply:
Done.  See [3f6b870](https://github.com/paketo-buildpacks/rfcs/pull/71/commits/3f6b870691c500a8d1e6201a2f990ea9761485e5)

Reply:
Thanks for fleshing that out. This does clarify things.